File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/r3f-require-uv-for-texture-map.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getAuthoritativeJsxAttribute } from "../../utils/get-authoritative-jsx-attribute.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { resolveJsxElementType } from "../../utils/resolve-jsx-element-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { UV_ATTRIBUTE_NAMES, UV_MAPPED_MATERIAL_CONSTRUCTOR_NAMES } from "./constants.js";
import { getActiveR3fMaterialTexturePropertyNames } from "./utils/get-active-r3f-material-texture-property-names.js";
import { getClosedR3fBufferGeometryAttributes } from "./utils/get-closed-r3f-buffer-geometry-attributes.js";
import { getR3fConstructorName } from "./utils/get-r3f-constructor-name.js";
import { getR3fSurfaceVisibility } from "./utils/get-r3f-surface-visibility.js";
import { hasR3fRuntimeImport } from "./utils/has-r3f-runtime-import.js";
import { isR3fHostIntrinsic } from "./utils/is-r3f-host-intrinsic.js";

export const r3fRequireUvForTextureMap = defineRule({
  id: "r3f-require-uv-for-texture-map",
  title: "Mapped R3F mesh geometry has no UVs",
  category: "Correctness",
  tags: ["react-jsx-only"],
  severity: "error",
  recommendation:
    "Attach a uv, uv1, uv2, or uv3 BufferAttribute when a built-in mesh material samples a texture map",
  create: (context: RuleContext) => {
    let importsReactThreeFiber = false;
    return {
      Program(node: EsTreeNodeOfType<"Program">) {
        importsReactThreeFiber = hasR3fRuntimeImport(node, context.scopes);
      },
      JSXElement(node: EsTreeNodeOfType<"JSXElement">) {
        const mesh = node.openingElement;
        if (
          !importsReactThreeFiber ||
          resolveJsxElementType(mesh) !== "mesh" ||
          !isR3fHostIntrinsic(mesh) ||
          mesh.attributes.some((attribute) => isNodeOfType(attribute, "JSXSpreadAttribute")) ||
          getAuthoritativeJsxAttribute(mesh.attributes, "geometry") ||
          getAuthoritativeJsxAttribute(mesh.attributes, "material")
        ) {
          return;
        }
        const elementChildren = node.children.filter((child) => isNodeOfType(child, "JSXElement"));
        const geometryChildren = elementChildren.filter(
          (child) => resolveJsxElementType(child.openingElement) === "bufferGeometry",
        );
        const mappedMaterialChildren = elementChildren.filter((child) => {
          const elementType = resolveJsxElementType(child.openingElement);
          return Boolean(
            elementType &&
            UV_MAPPED_MATERIAL_CONSTRUCTOR_NAMES.has(getR3fConstructorName(elementType)) &&
            !child.openingElement.attributes.some((attribute) =>
              isNodeOfType(attribute, "JSXSpreadAttribute"),
            ) &&
            getActiveR3fMaterialTexturePropertyNames(
              child.openingElement,
              getR3fConstructorName(elementType),
            ).size > 0 &&
            !getAuthoritativeJsxAttribute(child.openingElement.attributes, "attach"),
          );
        });
        if (geometryChildren.length !== 1 || mappedMaterialChildren.length !== 1) return;
        const hasUnknownChild = elementChildren.some(
          (child) => child !== geometryChildren[0] && child !== mappedMaterialChildren[0],
        );
        if (hasUnknownChild) return;
        const geometry = geometryChildren[0];
        const material = mappedMaterialChildren[0];
        if (
          !geometry ||
          !material ||
          !isR3fHostIntrinsic(geometry.openingElement) ||
          !isR3fHostIntrinsic(material.openingElement)
        ) {
          return;
        }
        const attributes = getClosedR3fBufferGeometryAttributes(geometry, context.scopes);
        if (
          !attributes.isComplete ||
          !attributes.attributeNames.has("position") ||
          [...UV_ATTRIBUTE_NAMES].some((attributeName) =>
            attributes.attributeNames.has(attributeName),
          )
        ) {
          return;
        }
        const materialType = resolveJsxElementType(material.openingElement);
        if (!materialType) return;
        if (getR3fSurfaceVisibility(node, material.openingElement, context) !== true) return;
        const texturePropertyNames = getActiveR3fMaterialTexturePropertyNames(
          material.openingElement,
          getR3fConstructorName(materialType),
        );
        context.report({
          node: geometry.openingElement,
          message: `${resolveJsxElementType(material.openingElement) ?? "This material"} samples ${[...texturePropertyNames].join(", ")}, but this custom bufferGeometry defines positions without any UV attribute`,
        });
      },
    };
  },
});

File: native/oxlint/rules/r3f-require-uv-for-texture-map.rs
use oxc_ast::{
    AstKind,
    ast::{JSXChild, JSXOpeningElement},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_span::GetSpan;

use crate::{
    context::{ContextHost, LintContext},
    rule::{Rule, RuleCategory, RuleInfo, RuleMeta},
};

const UV_ATTRIBUTE_NAMES: [&str; 4] = ["uv", "uv1", "uv2", "uv3"];

#[derive(Debug, Default, Clone)]
pub struct R3FRequireUvForTextureMap;

impl RuleMeta for R3FRequireUvForTextureMap {
    const NAME: &'static str = "r3f-require-uv-for-texture-map";
    const PLUGIN: &'static str = "react_doctor_native";
    const CATEGORY: RuleCategory = RuleCategory::Correctness;
    const VERSION: &'static str = "0.1.0";
    const INFO: RuleInfo = RuleInfo {
        short_description: "Require UVs for texture-mapped React Three Fiber geometry.",
    };
}

impl Rule for R3FRequireUvForTextureMap {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        ctx.source_type().is_jsx()
    }

    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        if file_is_non_react_jsx_dialect(ctx) || !has_r3f_runtime_import(ctx) {
            return;
        }
        let mut analysis = None;
        for node in ctx.nodes().iter() {
            let AstKind::JSXElement(mesh) = node.kind() else {
                continue;
            };
            if resolve_jsx_element_type(&mesh.opening_element, ctx)
                .is_none_or(|(element_type, _)| element_type != "mesh")
                || !is_r3f_host_intrinsic(&mesh.opening_element, ctx)
                || has_any_jsx_spread_attribute(&mesh.opening_element)
                || get_authoritative_jsx_attribute(&mesh.opening_element, "geometry", true)
                    .is_some()
                || get_authoritative_jsx_attribute(&mesh.opening_element, "material", true)
                    .is_some()
            {
                continue;
            }

            let element_children = mesh
                .children
                .iter()
                .filter_map(|child| match child {
                    JSXChild::Element(element) => Some(element.as_ref()),
                    _ => None,
                })
                .collect::<Vec<_>>();
            let geometry_children = element_children
                .iter()
                .copied()
                .filter(|element| {
                    resolve_jsx_element_type(&element.opening_element, ctx)
                        .is_some_and(|(element_type, _)| element_type == "bufferGeometry")
                })
                .collect::<Vec<_>>();
            let mapped_material_children = element_children
                .iter()
                .copied()
                .filter(|element| is_active_uv_mapped_material(&element.opening_element, ctx))
                .collect::<Vec<_>>();
            if geometry_children.len() != 1 || mapped_material_children.len() != 1 {
                continue;
            }
            let geometry = geometry_children[0];
            let material = mapped_material_children[0];
            if element_children.iter().any(|element| {
                !std::ptr::eq(*element, geometry) && !std::ptr::eq(*element, material)
            }) || !is_r3f_host_intrinsic(&geometry.opening_element, ctx)
                || !is_r3f_host_intrinsic(&material.opening_element, ctx)
            {
                continue;
            }

            let attributes = get_closed_r3f_buffer_geometry_attributes(geometry, ctx);
            if !attributes.is_complete
                || !attributes.attribute_names.contains("position")
                || UV_ATTRIBUTE_NAMES
                    .iter()
                    .any(|attribute_name| attributes.attribute_names.contains(*attribute_name))
            {
                continue;
            }
            let Some((material_type, _)) = resolve_jsx_element_type(&material.opening_element, ctx)
            else {
                continue;
            };
            if get_r3f_surface_visibility(
                mesh,
                &material.opening_element,
                node.id(),
                &mut analysis,
                ctx,
            ) != Some(true)
            {
                continue;
            }
            let texture_property_names = get_active_r3f_material_texture_property_names(
                &material.opening_element,
                &r3f_constructor_name(material_type),
            );
            ctx.diagnostic(
                OxcDiagnostic::error(format!(
                    "{material_type} samples {}, but this custom bufferGeometry defines positions without any UV attribute",
                    texture_property_names.join(", ")
                ))
                .with_label(geometry.opening_element.span),
            );
        }
    }
}

fn is_active_uv_mapped_material<'a>(
    opening_element: &JSXOpeningElement<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    let Some((element_type, _)) = resolve_jsx_element_type(opening_element, ctx) else {
        return false;
    };
    !has_any_jsx_spread_attribute(opening_element)
        && !get_active_r3f_material_texture_property_names(
            opening_element,
            &r3f_constructor_name(element_type),
        )
        .is_empty()
        && get_authoritative_jsx_attribute(opening_element, "attach", true).is_none()
}
